azblob/zt_url_block_blob_test.go

```diff
@@ -144,6 +144,77 @@ func (s *aztestsSuite) TestStageBlockFromURL(c *chk.C) {
+	testSize := 8 * 1024 * 1024 // 8MB
+	r, sourceData := getRandomDataAndReader(testSize)
+	sourceDataMD5Value := md5.Sum(sourceData)
+	ctx := context.Background() // Use default Background context
+	srcBlob := container.NewBlockBlobURL(generateBlobName())
+	destBlob := container.NewBlockBlobURL(generateBlobName())
+
+	// Prepare source blob for copy.
+	uploadSrcResp, err := srcBlob.Upload(ctx, r, azblob.BlobHTTPHeaders{}, azblob.Metadata{}, azblob.BlobAccessConditions{})
+	c.Assert(err, chk.IsNil)
+	c.Assert(uploadSrcResp.Response().StatusCode, chk.Equals, 201)
+
+	// Get source blob URL with SAS for StageFromURL.
+	srcBlobParts := azblob.NewBlobURLParts(srcBlob.URL())
+
+	srcBlobParts.SAS, err = azblob.BlobSASSignatureValues{
+		Protocol:      azblob.SASProtocolHTTPS,              // Users MUST use HTTPS (not HTTP)
+		ExpiryTime:    time.Now().UTC().Add(48 * time.Hour), // 48-hours before expiration
+		ContainerName: srcBlobParts.ContainerName,
+		BlobName:      srcBlobParts.BlobName,
+		Permissions:   azblob.BlobSASPermissions{Read: true}.String(),
+	}.NewSASQueryParameters(credential)
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	srcBlobURLWithSAS := srcBlobParts.URL()
+
+	// Invoke copy blob from URL.
+	resp, err := destBlob.CopyFromURL(ctx, srcBlobURLWithSAS, azblob.Metadata{"foo": "bar"}, azblob.ModifiedAccessConditions{}, azblob.BlobAccessConditions{}, sourceDataMD5Value[:])
+	c.Assert(err, chk.IsNil)
+	c.Assert(resp.Response().StatusCode, chk.Equals, 202)
+	c.Assert(resp.ETag(), chk.Not(chk.Equals), "")
+	c.Assert(resp.RequestID(), chk.Not(chk.Equals), "")
+	c.Assert(resp.Version(), chk.Not(chk.Equals), "")
+	c.Assert(resp.Date().IsZero(), chk.Equals, false)
+	c.Assert(resp.CopyID(), chk.Not(chk.Equals), "")
+	c.Assert(resp.ContentMD5(), chk.DeepEquals, sourceDataMD5Value[:])
+	c.Assert(string(resp.CopyStatus()), chk.DeepEquals, "success")
+
+	// Check data integrity through downloading.
+	downloadResp, err := destBlob.BlobURL.Download(ctx, 0, azblob.CountToEnd, azblob.BlobAccessConditions{}, false)
+	c.Assert(err, chk.IsNil)
+	destData, err := ioutil.ReadAll(downloadResp.Body(azblob.RetryReaderOptions{}))
+	c.Assert(err, chk.IsNil)
+	c.Assert(destData, chk.DeepEquals, sourceData)
+
+	// Make sure the metadata got copied over
+	c.Assert(len(downloadResp.NewMetadata()), chk.Equals, 1)
+
+	// Edge case 1: Provide bad MD5 and make sure the copy fails
+	_, badMD5 := getRandomDataAndReader(16)
+	_, err = destBlob.CopyFromURL(ctx, srcBlobURLWithSAS, azblob.Metadata{}, azblob.ModifiedAccessConditions{}, azblob.BlobAccessConditions{}, badMD5)
+	c.Assert(err, chk.NotNil)
+
+	// Edge case 2: Not providing any source MD5 should see the CRC getting returned instead
+	resp, err = destBlob.CopyFromURL(ctx, srcBlobURLWithSAS, azblob.Metadata{}, azblob.ModifiedAccessConditions{}, azblob.BlobAccessConditions{}, nil)
+	c.Assert(err, chk.IsNil)
+	c.Assert(resp.Response().StatusCode, chk.Equals, 202)
+	c.Assert(resp.XMsContentCrc64(), chk.Not(chk.Equals), "")
+}
```

Comment:
I would ensure we have decent logging for when a user tries to copy a blob that's too large, to ensure we're not just handing back a raw 400 with no real info.

Reply:
hmm that feel too much like testing the service though...